ccl/backupccl,kv/bulk: add tracing to restore processor and sst batcher

rhu713 opened the pull request:
Enable job tracing in restore jobs. Use bulk aggregator to collect throughput
and SST Batcher flush stats as part of tracing in restore.

Release justification: low risk, high benefit changes to existing functionality

Release note: None

Commit:
ccl/backupccl,kv/bulk: add tracing to restore processor, sst batcher, buf adder

Enable job tracing in restore jobs. Use bulk aggregator to collect what was
previously IngestionPerformanceStats from buffering adders and SST batchers.
Aggregators are created for each restore processor and for each worker inside a
processor.

Release justification: low risk, high benefit changes to existing functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/87013)
<!-- Reviewable:end -->

adityamaru commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -402,6 +407,9 @@ func (rd *restoreDataProcessor) processRestoreSpanEntry(
 	iter := sst.iter
 	defer sst.cleanup()
 
+	ctx, agg := bulkutil.MakeTracingAggregatorWithSpan(ctx, "SSTBatcher-aggregator", evalCtx.Tracer)
```
hmm do we want to create a new aggregator for every restore span entry we process? I don't think this span will be very long-lived and so it will only be aggregating information for each sst we're ingesting and might not even be easy to view in tracez unless we capture a snapshot that catches this span in action.

What do you think about moving this to a higher level maybe outside the `for loop` where `processRestoreSpanEntry` is called in `runRestoreWorkers`. Lets create an aggregator per worker we spawn? This will be longer lived than this span and will aggregate information across SSTs ingested by the worker.

rhu713 replied:
done, use an aggregator for each worker instead of each span

adityamaru commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -160,6 +163,8 @@ func (rd *restoreDataProcessor) Start(ctx context.Context) {
 	rd.input.Start(ctx)
 
 	ctx, cancel := context.WithCancel(ctx)
+	ctx, rd.agg = bulkutil.MakeTracingAggregatorWithSpan(ctx, fmt.Sprintf("%s-aggregator", restoreDataProcName), rd.EvalCtx.Tracer)
```
If we have an aggregator per restore worker as discussed below, I guess this span gives us a view of all the ingestion occurring across our workers? It could be useful, I think it'd help if we ran a longish restore and actually saw these tags in action. Can you maybe run a quick restore an see this playout on `tracez`?

rhu713 replied:
Yes, it does give an overall view of the ingestion among all workers

adityamaru commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -876,3 +890,23 @@ func createSplitSSTable(
 	return left, right, nil
 }
+
+func (b *SSTBatcher) reportFlushStats(size sz, reason int, lastFlush, afterFlush time.Time) {
```
What do you think about reworking this to emit the `ingestionStats` @dt has already fleshed out in the past? It holds a lot of valuable information that we can surface (maybe after some pre-processing) that is specific to this SSTBatcher that can be emitted every time it flushes or at some pre-defined interval. This would then get aggregated and displayed by the restore data processor.

rhu713 replied:
done, used ingestion stats as the exported event instead of a new event.

adityamaru commented on pkg/ccl/backupccl/backuppb/backup.proto:
```diff
@@ -206,3 +206,18 @@ message ExportStats {
 
+// SSTBatcherStats is a message containing information about the batches created
+// by the SST Batcher.
+message SSTBatcherStats {
```
Noted below but what if we turn `ingestionStats` from a struct into a protobuf and emit that instead? This way we avoid two parallel tracks of collecting stats. We could process those stats to pick out the useful pieces of information that we wish to display but its easier to reason about a single source of truth.

rhu713 reviewed (commented):
@adityamaru do you mind giving this another look? I'm not sure why CI is not passing but i think the changes should be ready for a look.

adityamaru commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -585,18 +598,31 @@ func (b *SSTBatcher) doFlush(ctx context.Context, reason int) error {
 		}
 		b.mu.Lock()
 		summary.DataSize += int64(size)
+		currentBatchStatsCopy.DataSize += int64(size)
```
Is it safe to drop the atomic add because we're holding on to the b.mu lock while updating these stats?

rhu713 replied:
This should be okay to modify because of the same copy reason as above.

adityamaru replied:
Same comment as below, I agree we're operating on a copy but isn't that copy shared by concurrent invocations of `fn`?

adityamaru replied:
Ohh I think i understand. Even if we have concurrent invocation of `fn` across flushes, each one will have its own copy. If this is true, maybe a quick comment to this effect?

adityamaru commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -585,18 +598,31 @@ func (b *SSTBatcher) doFlush(ctx context.Context, reason int) error {
+		b.lastFlush = afterFlush
 		b.mu.Unlock()
+
+		if b.tracingSpan != nil {
```
Does this need to be in the mutex too?

rhu713 replied:
no, because the stats copy is what's being logged.

adityamaru replied:
Right, but isn't b.tracingSpan being called from concurrent invocation of `fn`?

adityamaru reviewed (changes requested):
im slightly confused by the direct manipulation of `totalStats` in parts of the sst batcher. I was hoping that the only time we modify totalStats would be when we combine the currentBatchStats into the total stats, everywhere else we'd modify the batch stats. As things stand today I believe the structured event doesn't render these fields that are only modified on `totalStats`?

Do you think it'll be easier to have a `totalStats` and a `currentStats` instead of trying to do a per-batch stats? This way we would only always manipulate `currentStats` and everytime we go to emit the structured event (which is on flush), we'd reset the currentStats after combining them into the total stats.

rhu713 commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -725,18 +755,16 @@ func (b *SSTBatcher) addSSTable(
 				sendTime := timeutil.Since(beforeSend)
 
-				atomic.AddInt64(&b.stats.sendWaitAtomic, int64(sendTime))
+				ingestionPerformanceStats.SendWait += sendTime
```
I think because this is done on the copy of `currentStats` as opposed to a shared stats variable before, and there's only 1 goroutine modifying this copy, that anything in the `ingestionPerformanceStats` is okay to modify.

adityamaru replied:
So `fn` can be called concurrently several times IIUC. We're passing in a pointer to the copy of currentStats and each of the concurrent invocations will mutate that same pointer. Is that not susceptible to concurrent writes?

adityamaru replied:
I think I resolved my own question above, if you can just confirm my understanding. Thanks!

rhu713 replied:
correct, this `ingestionPerformanceStats` pointer is the pointer to the copy of currentStats, and is only modified by 1 invocation of `fn`

rhu713 commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -648,8 +674,12 @@ func (b *SSTBatcher) addSSTable(
 ) error {
 	sendStart := timeutil.Now()
+	if ingestionPerformanceStats == nil {
+		panic("ingestionPerformanceStats should not be nil")
```
done

rhu713 commented on pkg/ccl/backupccl/backuppb/backup.go:
```diff
@@ -179,11 +191,219 @@ func (e *ExportStats) Combine(other bulk.TracingAggregatorEvent) {
+}
+
+// LogTimings logs the timing ingestion stats.
+func (s IngestionPerformanceStats) LogTimings(ctx context.Context, name, action string) {
```
done, used pointer receiver

rhu713 commented:
> im slightly confused by the direct manipulation of totalStats in parts of the sst batcher

Yeah good call, I was a little confused by what I had there too. I changed a few things so hopefully it's more clear on what needs to be guarded by a mutex.
1. I put `totalStats` behind the batcher mutex as I think it's the only object that could be read and modified concurrently by concurrent calls of `fn`. Most of the reads that are for logging purposes are still read without acquiring the mutex.
2. I changed all of the modifications of `totalStats` to `currentStats` so the stats would be emitted on flush.

adityamaru commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -139,7 +140,16 @@ type SSTBatcher struct {
+	// span tracks the total span into which this batcher has flushed. It is
+	// only maintained if log.V(1), so if vmodule is upped mid-ingest it may be
+	// incomplete.
+	span roachpb.Span
```
we could move this into the IngestionStats proto?

adityamaru commented on pkg/kv/bulk/buffering_adder.go:
```diff
@@ -136,12 +137,12 @@ func (b *BufferingAdder) SetOnFlush(fn func(summary roachpb.BulkOpSummary)) {
 	if log.V(1) {
-		if b.sink.stats.bufferFlushes > 0 {
-			b.sink.stats.LogTimings(ctx, b.name, "closing")
+		if b.sink.mu.totalStats.BufferFlushes > 0 {
```
We're not locking the mutex before reading here?

adityamaru commented on pkg/kv/bulk/buffering_adder.go:
```diff
@@ -287,18 +291,22 @@ func (b *BufferingAdder) doFlush(ctx context.Context, forSize bool) error {
-		dueToSplits := b.sink.stats.batchesDueToRange - before.batchesDueToRange
-		dueToSize := b.sink.stats.batchesDueToRange - before.batchesDueToRange
+		afterStats := b.sink.mu.totalStats.Identity().(*backuppb.IngestionPerformanceStats)
+		afterStats.Combine(&b.sink.mu.totalStats)
```
If beforeStats = totalStats + curStats before flush, and afterStats = totalStats + curStats after flush, isn't it adequate to do curStats after flush - curStats before flush?

adityamaru commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -171,7 +181,11 @@ type SSTBatcher struct {
 
 		maxWriteTS hlc.Timestamp
 		totalRows  roachpb.BulkOpSummary
+		totalStats backuppb.IngestionPerformanceStats
```
nit: Can you add a comment here about when totalStats is aggregated in the batcher

adityamaru commented on pkg/kv/bulk/sst_batcher.go:
```diff
@@ -585,18 +604,31 @@ func (b *SSTBatcher) doFlush(ctx context.Context, reason int) error {
+		currentBatchStatsCopy.BatchWait += afterFlush.Sub(beforeFlush)
+		currentBatchStatsCopy.Duration = afterFlush.Sub(b.lastFlush)
+		b.mu.totalStats.Combine(currentBatchStatsCopy)
+		b.lastFlush = afterFlush
```
Should we put `lastFlush` in `mu` given that concurrent fns can mutate it?

rhu713 replied:
done, put `lastFlush` in `mu`

adityamaru commented on pkg/ccl/backupccl/backuppb/backup.go:
```diff
@@ -179,11 +191,219 @@ func (e *ExportStats) Combine(other bulk.TracingAggregatorEvent) {
+}
+
+// Render implements the TracingAggregatorEvent interface.
+func (s *IngestionPerformanceStats) Render() []attribute.KeyValue {
```
I love it! Can't wait to see this in action.

adityamaru reviewed (changes requested):
Almost there, do you think we could add a sanity check test like the maybe in `tracing_aggregator_test.go` that records an IngestionStats event a few times and makes sure its getting combined as expected? Maybe you can modify the existing test that uses `ExportRequestStats`.

adityamaru commented:
cc'ing @dt in case he is interested in taking a peek

rhu713 commented:
> add a sanity check test like the maybe in tracing_aggregator_test.go

Added the sanity test, good call!

adityamaru commented on pkg/ccl/backupccl/backuppb/backup.go:
```diff
@@ -179,11 +191,220 @@ func (e *ExportStats) Combine(other bulk.TracingAggregatorEvent) {
+				Value: attribute.Int64Value(s.BatchesDueToRange),
+			},
+			attribute.KeyValue{
+				Key:   "split_retires",
```
nit: lets define all strings as top level consts? Or inline all the strings for consistency.

adityamaru reviewed (approved):
LGTM with two questions. Given all the additional mutex locking/unlocking that only happens under vmodule, can you run a restore with vmodule set to 3 or 4 on buffering_adder/sst_batcher. I'm afraid of a deadlock we'd only see with increased verbosity which is not the mode we test under every night. Maybe we should do that for our roachtests in a follow up.

rhu713 commented:
bors r+
